### ISSUE TITLE
Need to pass Security Labels and potentially USERNAMESPACE mappings to volumes.

### ISSUE THREAD

**Issue by rhatdan** [docker/docker#20859]
We have been playing around with a lvm volume plugin for docker.  One problem we are having is how to setup the security constraints of the volume.

https://github.com/shishir-a412ed/docker-lvm-plugin

The way this volume is working currently is during docker create the volume gets created without any SELinux labels associated with it and without any User Namespace.  Then when you run `docker run` it will get mounted into the container with a label that is blocked by SELinux and if you were running with User Namespace, owned by UID=0, which would not be accessible to the container.

We have looked at making changes to docker and doing a relabel during setupMounts, but this means every time you started the container, you would be running recursive changes of SELinux Labels and potentially UID/GID parings.  We want to only do this during container creation.  Or in the case of lvm we could mount the volume with context='SELINUX_LABEL".  But none of this information is available to the volume plugin.

Just wanted to open a discussion on this issue.

**Comment by rhatdan** [comment]
@cpuguy83 I guess you would be the lead on this discussion.

**Comment by cpuguy83** [comment]
@rhatdan Does the userns example not work on 1.10?

**Comment by cpuguy83** [comment]
In 1.10 the content of the container fs should be copied into the volume, taking the uid/gid with it.

**Comment by rhatdan** [comment]
cpuguy83, if I used one of your new "mount flags, I believe I could set the UID to match, but I don't see how the top level pid would get created with the correct UID.  We could also use the SELinux labels on the -o option to set up the labels to be shared but I can't set them to be private to the container, since I don't know the MCS Label until the container has been created.

In the future if I we added the ability to specify the UID/GID Usernamespace fields on the docker run command, how would the UID/GID of the ownership pid of the volume be set?  

Currently we are setting local volumes during the creation stage, but if the actual volumes are not created/mounted until the mount stage, I don't see them being set correctly.

**Comment by thaJeztah** [comment]
@cpuguy83 @rhatdan what's the status on this one? Is this still needed?

**Comment by rhatdan** [comment]
I am not sure

**Comment by jmccormick2001** [comment]
this bug effects me, would be great to see it fixed..thanks.  running 1.12 in enforcing mode, I have to run the container as privledged to get around this using the local volume driver.

**Comment by thaJeztah** [comment]
Is this related to https://github.com/docker/docker/pull/23024 ?

**Comment by rhatdan** [comment]
Yes this fixes this issue at least as far as SELinux is concerned.
